extensions/gitpod/src/remoteConnector.ts

```diff
@@ -379,63 +380,149 @@ export default class LocalApp extends Disposable {
+		if (action === cancel) {
+			return;
+		}
+
+		await vscode.commands.executeCommand('extension.open', 'ms-vscode-remote.remote-ssh');
+		await vscode.commands.executeCommand('workbench.extensions.installExtension', 'ms-vscode-remote.remote-ssh');
+	}
+
 	public async handleUri(uri: vscode.Uri) {
-		if (uri.path === LocalApp.AUTH_COMPLETE_PATH) {
+		if (uri.path === RemoteConnector.AUTH_COMPLETE_PATH) {
 			this.logger.info('auth completed');
 			return;
 		}
-		this.logger.info('open workspace window: ' + uri.toString());
+
+		await this.ensureRemoteSSHExtInstalled();
+
+		this.logger.info('Opening remote workspace', uri.toString());
+
 		const params: SSHConnectionParams = JSON.parse(uri.query);
-		let resolvedConfig: LocalAppConfig | undefined;
+		let localAppLogPath: string | undefined;
+		let localAppSSHConfigPath: string | undefined;
 		try {
-			await vscode.window.withProgress({
-				location: vscode.ProgressLocation.Notification,
-				cancellable: true,
-				title: `Connecting to Gitpod workspace: ${params.workspaceId}`
-			}, async (_, token) => {
-				const connection = await this.withLocalApp(params.gitpodHost, (client, config) => {
-					resolvedConfig = config;
-					const request = new ResolveSSHConnectionRequest();
-					request.setInstanceId(params.instanceId);
-					request.setWorkspaceId(params.workspaceId);
-					return new Promise<ResolveSSHConnectionResponse>((resolve, reject) =>
-						client.resolveSSHConnection(request, (e, r) => r ? resolve(r) : reject(e))
-					);
-				}, token);
-
-				const config = vscode.workspace.getConfiguration('remote.SSH');
-				const defaultExtensions = config.get<string[]>('defaultExtensions') || [];
-				if (defaultExtensions.indexOf('gitpod.gitpod-remote-ssh') === -1) {
-					defaultExtensions.unshift('gitpod.gitpod-remote-ssh');
-					await config.update('defaultExtensions', defaultExtensions, vscode.ConfigurationTarget.Global);
-				}
-				// TODO(ak) notify a user about config file changes?
-				const gitpodConfigFile = connection.getConfigFile();
-				const currentConfigFile = config.get<string>('configFile');
-				if (currentConfigFile === gitpodConfigFile) {
-					// invalidate cached SSH targets from the current config file
-					await config.update('configFile', undefined, vscode.ConfigurationTarget.Global);
-				}
-				await config.update('configFile', gitpodConfigFile, vscode.ConfigurationTarget.Global);
-				// TODO(ak) ensure that vscode.ssh-remote is installed
-				await vscode.commands.executeCommand('vscode.openFolder', vscode.Uri.parse(`vscode-remote://ssh-remote+${connection.getHost()}${uri.path || '/'}`), {
-					forceNewWindow: true
-				});
-			});
+			let sshDestination = await this.getWorkspaceSSHDestination(params.workspaceId);
+
+			const usingSSHGateway = !!sshDestination;
+			if (!usingSSHGateway) {
+				const localAppDestData = await this.getWorkspaceLocalAppSSHDestination(params);
```

Comment:
we should warn here about usage of deprecated connection mode and that installation of gitpod host from SSHConnectionParams should be upgraded to enable direct SSH access.

Reply:
a link to docs: https://github.com/gitpod-io/gitpod/blob/main/install/installer/docs/workspace-ssh-access.md